Metrics Endpoint Causing Errors

This starts happening at an unknown time. So far I have been unable to repro it locally. It may be load related or memory related. Needs more investigation.

Internal server error, for (GET) [/metrics] ->

> play.api.UnexpectedException: Unexpected exception[ConcurrentModificationException: null]	 
>     at play.api.http.HttpErrorHandlerExceptions$.throwableToUsefulException(HttpErrorHandler.scala:384)	 
>     at play.api.http.HttpErrorHandlerExceptions.throwableToUsefulException(HttpErrorHandler.scala)	 
>     at play.http.DefaultHttpErrorHandler.throwableToUsefulException(DefaultHttpErrorHandler.java:245)	 
>     at play.http.DefaultHttpErrorHandler.onServerError(DefaultHttpErrorHandler.java:181)	 
>     at controllers.ErrorHandler.onServerError(ErrorHandler.java:127)	 
>     at play.core.j.JavaHttpErrorHandlerAdapter.onServerError(JavaHttpErrorHandlerAdapter.scala:33)	 
>     at play.core.server.PekkoHttpServer$$anonfun$invokeAction$1$1.applyOrElse(PekkoHttpServer.scala:482)	 
>     at play.core.server.PekkoHttpServer$$anonfun$invokeAction$1$1.applyOrElse(PekkoHttpServer.scala:474)	 
>     at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:490)	 
>     at org.apache.pekko.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:73)	 
>     at org.apache.pekko.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:110)	 
>     at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)	 
>     at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:94)	 
>     at org.apache.pekko.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:110)	 
>     at org.apache.pekko.dispatch.TaskInvocation.run(AbstractDispatcher.scala:59)	 
>     at org.apache.pekko.dispatch.ForkJoinExecutorConfigurator$PekkoForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:57)	 
>     at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)	 
>     at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)	 
>     at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)	 
>     at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source) 
> 
> play.api.UnexpectedException: Unexpected exception[ConcurrentModificationException: null] 
>     at play.api.http.HttpErrorHandlerExceptions$.throwableToUsefulException(HttpErrorHandler.scala:384) 
>     at play.api.http.HttpErrorHandlerExceptions.throwableToUsefulException(HttpErrorHandler.scala) 
>     at play.http.DefaultHttpErrorHandler.throwableToUsefulException(DefaultHttpErrorHandler.java:245) 
>     at play.http.DefaultHttpErrorHandler.onServerError(DefaultHttpErrorHandler.java:181) 
>     at controllers.ErrorHandler.onServerError(ErrorHandler.java:127) 
>     at play.core.j.JavaHttpErrorHandlerAdapter.onServerError(JavaHttpErrorHandlerAdapter.scala:33) 
>     at play.core.server.PekkoHttpServer$$anonfun$invokeAction$1$1.applyOrElse(PekkoHttpServer.scala:482) 
>     at play.core.server.PekkoHttpServer$$anonfun$invokeAction$1$1.applyOrElse(PekkoHttpServer.scala:474) 
>     at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:490) 
>     at org.apache.pekko.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:73) 
>     at org.apache.pekko.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:110) 
>     at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18) 
>     at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:94) 
>     at org.apache.pekko.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:110) 
>     at org.apache.pekko.dispatch.TaskInvocation.run(AbstractDispatcher.scala:59) 
>     at org.apache.pekko.dispatch.ForkJoinExecutorConfigurator$PekkoForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:57) 
>     at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source) 
>     at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source) 
>     at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source) 
>     at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
> 
> Caused by: java.util.ConcurrentModificationException: null	 
>     at java.base/java.util.ArrayList$Itr.checkForComodification(Unknown Source)	 
>     at java.base/java.util.ArrayList$Itr.next(Unknown Source)	 
>     at io.ebeaninternal.server.transaction.TransactionManager.visitMetrics(TransactionManager.java:424)	 
>     at io.ebeaninternal.server.core.DefaultServer.visitMetrics(DefaultServer.java:2094)	 
>     at io.ebeaninternal.server.core.DefaultMetaInfoManager.visitMetrics(DefaultMetaInfoManager.java:33)	 
>     at io.ebeaninternal.server.core.DefaultMetaInfoManager.visitBasic(DefaultMetaInfoManager.java:44)	 
>     at io.ebeaninternal.server.core.DefaultMetaInfoManager.collectMetrics(DefaultMetaInfoManager.java:38)	 
>     at controllers.monitoring.MetricsController.getMetrics(MetricsController.java:67)	 
>     at router.Routes$$anonfun$routes$1.$anonfun$applyOrElse$520(Routes.scala:5254)	 
>     at play.core.routing.HandlerInvokerFactory$$anon$6.resultCall(HandlerInvoker.scala:152)	 
>     at play.core.routing.HandlerInvokerFactory$JavaActionInvokerFactory$$anon$3$$anon$4$$anon$5.invocation(HandlerInvoker.scala:125)	 
>     at play.core.j.JavaAction$$anon$1.$anonfun$call$1(JavaAction.scala:128)	 
>     at play.api.mvc.BodyParser$.$anonfun$parseBody$4(Action.scala:241)	 
>     at scala.Option.getOrElse(Option.scala:201)	 
>     at play.api.mvc.BodyParser$.parseBody(Action.scala:239)	 
>     at play.core.j.JavaAction$$anon$1.call(JavaAction.scala:136)	 
>     at play.http.DefaultActionCreator$1.call(DefaultActionCreator.java:31)	 
>     at play.core.j.JavaAction.$anonfun$apply$8(JavaAction.scala:192)	 
>     at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:687)	 
>     at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:467)
> 

**Additional context**
Add any other context about the problem here.

--- a/server/app/controllers/monitoring/MetricsController.java
+++ b/server/app/controllers/monitoring/MetricsController.java
@@ -7,12 +7,14 @@ import com.typesafe.config.Config;
 import controllers.CiviFormController;
 import io.ebean.DB;
 import io.ebean.Database;
+import io.ebean.meta.ServerMetrics;
 import io.prometheus.client.CollectorRegistry;
 import io.prometheus.client.exporter.common.TextFormat;
-import java.io.IOException;
 import java.io.StringWriter;
-import java.io.UncheckedIOException;
+import java.util.ConcurrentModificationException;
 import javax.inject.Inject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 import play.mvc.Result;
 import repository.VersionRepository;
 import services.monitoring.MonitoringMetricCounters;
@@ -23,6 +25,7 @@ import services.monitoring.MonitoringMetricCounters;
  * customized to allow disabling via configuration flag.
  */
 public final class MetricsController extends CiviFormController {
+  private static final Logger LOGGER = LoggerFactory.getLogger(MetricsController.class);
 
   private final boolean metricsEnabled;
   private final CollectorRegistry collectorRegistry;
@@ -59,12 +62,8 @@ public final class MetricsController extends CiviFormController {
       return notFound();
     }
 
-    var writer = new StringWriter();
-
     try {
-      database
-          .metaInfo()
-          .collectMetrics()
+      collectMetrics()
           .queryMetrics()
           .forEach(
               metric -> {
@@ -97,11 +96,44 @@ public final class MetricsController extends CiviFormController {
                     .inc((double) metric.total());
               });
 
+      var writer = new StringWriter();
       TextFormat.write004(writer, collectorRegistry.metricFamilySamples());
-    } catch (IOException e) {
-      throw new UncheckedIOException(e);
+      return ok(writer.toString()).as(TextFormat.CONTENT_TYPE_004);
+    } catch (ConcurrentModificationException e) {
+      // This exception can often be reproduced if you start running the full browser test
+      // suite and then spam calling `/metrics` until it sometimes will throw.
+      //
+      // Basically this state can happen when calling `database.metaInfo().collectMetrics()`
+      // method and the metrics are either being updated or a lock can't be immediately gotten.
+      //
+      // I'm trying to avoid using synchronize to force a lock on the database object as I'm
+      // concerned about adversely affecting performance.
+      //
+      // Since the `/metrics` endpoint is called very frequently any data from a failed call
+      // gets rolled into the next successful call to `/metrics`.
+      var errorMsg =
+          """
+          Attempted to collect database metrics for Prometheus, but could not get a lock at this
+          time. This is safe to ignore. The metrics will be included on the next successful call.
+          """;
+
+      LOGGER.info(errorMsg.stripIndent(), e);
+    } catch (Exception e) {
+      LOGGER.error(e.getMessage());
     }
 
-    return ok(writer.toString()).as(TextFormat.CONTENT_TYPE_004);
+    return internalServerError();
+  }
+
+  /**
+   * Attempt to collect server metrics. Will make one extra attempt in the event of a
+   * ConcurrentModificationException
+   */
+  private ServerMetrics collectMetrics() {
+    try {
+      return database.metaInfo().collectMetrics();
+    } catch (ConcurrentModificationException e) {
+      return database.metaInfo().collectMetrics();
+    }
   }
 }

--- a/server/app/controllers/monitoring/MetricsController.java
+++ b/server/app/controllers/monitoring/MetricsController.java
@@ -100,14 +100,13 @@ public final class MetricsController extends CiviFormController {
       TextFormat.write004(writer, collectorRegistry.metricFamilySamples());
       return ok(writer.toString()).as(TextFormat.CONTENT_TYPE_004);
     } catch (ConcurrentModificationException e) {
-      // This exception can often be reproduced if you start running the full browser test
-      // suite and then spam calling `/metrics` until it sometimes will throw.
+      // This exception can often be triggered if you start running the full browser test
+      // suite and then spam calling `/metrics`.
       //
       // Basically this state can happen when calling `database.metaInfo().collectMetrics()`
       // method and the metrics are either being updated or a lock can't be immediately gotten.
       //
-      // I'm trying to avoid using synchronize to force a lock on the database object as I'm
-      // concerned about adversely affecting performance.
+      // Avoid using synchronize to force a lock on the database object as it may adversely affecting performance.
       //
       // Since the `/metrics` endpoint is called very frequently any data from a failed call
       // gets rolled into the next successful call to `/metrics`.

--- a/server/app/controllers/monitoring/MetricsController.java
+++ b/server/app/controllers/monitoring/MetricsController.java
@@ -7,7 +7,6 @@ import com.typesafe.config.Config;
 import controllers.CiviFormController;
 import io.ebean.DB;
 import io.ebean.Database;
-import io.ebean.meta.ServerMetrics;
 import io.prometheus.client.CollectorRegistry;
 import io.prometheus.client.exporter.common.TextFormat;
 import java.io.StringWriter;
@@ -63,7 +62,9 @@ public final class MetricsController extends CiviFormController {
     }
 
     try {
-      collectMetrics()
+      database
+          .metaInfo()
+          .collectMetrics()
           .queryMetrics()
           .forEach(
               metric -> {
@@ -106,7 +107,8 @@ public final class MetricsController extends CiviFormController {
       // Basically this state can happen when calling `database.metaInfo().collectMetrics()`
       // method and the metrics are either being updated or a lock can't be immediately gotten.
       //
-      // Avoid using synchronize to force a lock on the database object as it may adversely affecting performance.
+      // Avoid using synchronize to force a lock on the database object as it may adversely
+      // affecting performance.
       //
       // Since the `/metrics` endpoint is called very frequently any data from a failed call
       // gets rolled into the next successful call to `/metrics`.
@@ -122,17 +124,5 @@ public final class MetricsController extends CiviFormController {
     }
 
     return internalServerError();
-  }
-
-  /**
-   * Attempt to collect server metrics. Will make one extra attempt in the event of a
-   * ConcurrentModificationException
-   */
-  private ServerMetrics collectMetrics() {
-    try {
-      return database.metaInfo().collectMetrics();
-    } catch (ConcurrentModificationException e) {
-      return database.metaInfo().collectMetrics();
-    }
   }
 }

